### DISCUSSION TITLE
No unique rule?

### DISCUSSION THREAD

**Discussion by Tofandel** [Edwin-Luijten/strapi-encryptable-field#6] [General]
Strapi fields normally have a few validation options

![image](https://user-images.githubusercontent.com/6115458/228303692-242bc496-edd0-40c0-9c13-daf1b5b4db36.png)
One of which is the `unique` rule

This plugin doesn't have it
![image](https://user-images.githubusercontent.com/6115458/228303917-5cbf3551-6cc9-4133-bd34-d3141ff0436b.png)

Is it possible to add it?

**Comment by Edwin-Luijten** [comment]
Hello Tofandel.

The reason there is no unique rule is that an encrypted value is always unique. To fix this we need to be able to add a new field or even more if you want certain ways to look it up (first 3 letters of a name for example). For each of those you'll want to create a hmac.

I started working on a specific field type for this but I did hit a wall. When I have a bit more time i'll put my teeth in it again.

**Comment by Tofandel** [comment]
I see because of the IV, so yes the field would need another column with it's hashed value for the unique rule to work
But that means the column need to be generated on the fly when the unique rule is activated or not

Another possibility is to add the hmac directly into the field itself, it can even be used to make sure the encrypted value was not modified (authenticated encryption) making the encryption more secure and then be used for unicity as well

**Comment by alexandrepaiva-dev** [comment]
Hey guys,

Is there any update for this?

Ideally we should encrypt fields like email/username as they are sensitive data (GDPR) but at the same time they should be unique fields.

How can we achieve this?
